### ISSUE TITLE
Support more subscript symbols

### ISSUE THREAD

**Issue by liwt31** [dgasmith/opt_einsum#67]
The project now supports more than 52 subscripts, but only with integers, and it can quickly get confusing using so many integers. My suggestion is to support this kind of syntax:
```
arr1 = np.zeros((2, 3))
sub1 = '[left edge][bond]'
arr2 = np.zeros((3, 2))
sub2 = '[bond][right edge]'
out = '[left edge][right edge]'
opt_einsum.contract(arr1, sub1, arr2, sub2, out)
```

I know that it's equivalent with:
```
arr1 = np.zeros((2, 3))
sub1 = [0, 1]
arr2 = np.zeros((3, 2))
sub2 = [1, 2]
out = [0, 2]
```

However, IMHO, when doing contraction with lots of tensor which have different meanings, it'll be nice to give them a proper name.

I'm more than happy to prepare a PR for this, I think it's not extemely difficult.

**Comment by jcmgray** [comment]
Yes that would be how to specify the output indices - for other contraction syntaxes this doesn't need to specified, its just the alphabetic order of any uncontracted (single) indices, I imagine alphabetic- or input-order would be a good default here, as well.

> I am deeply impressed by the quimb.tensor module. Have you ever thought about separating it as an individual package?

Thanks! Yes it occurred to me but I haven't had the time or need to decouple it quite yet..

**Comment by dgasmith** [comment]
@jcmgray Should we consider moving the edge/bond syntax into `opt_einsum` or leave it elsewhere? I would be a bit concerned that a third input form on `contact` could become unwieldy as we have the above mentioned integer input as well.

**Comment by jcmgray** [comment]
Yes I am in two minds too a bit. One thing that is possible is just to broaden the current case to anything hashable (which includes ``ints``) so it would not be two separate input forms. Here is useful snippet that could replace the current logic:

```python
>>> symbols = (oe.get_symbol(i) for i in itertools.count())
>>> symbol_map= defaultdict(lambda: next(symbols))
>>> symbol_map['left'] 
'a'
>>> symbol_map['right]
'b'
>>> symbol_map['left']
'a'
```
then you could replace the line
```python
subscripts += get_symbol(s)
```
just with
```python
subscripts += symbol_map[s]
```
at basically no cost.

**Comment by dgasmith** [comment]
Closed by #68.
